File: module_composer/src/main/java/app/melon/composer/ui/CheckView.java
package app.melon.composer.ui;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RadialGradient;
import android.graphics.Rect;
import android.graphics.Shader;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

import androidx.core.content.ContextCompat;

import androidx.core.content.res.ResourcesCompat;

import app.melon.composer.R;

public class CheckView extends View {

    private static final float STROKE_WIDTH = 3.0f; // dp
    private static final float SHADOW_WIDTH = 6.0f; // dp
    private static final int SIZE = 48; // dp
    private static final float STROKE_RADIUS = 11.5f; // dp
    private static final float BG_RADIUS = 11.0f; // dp
    private static final int CONTENT_SIZE = 16; // dp
    
    private boolean mChecked;

    private Paint mStrokePaint;
    private Paint mBackgroundPaint;
    private TextPaint mTextPaint;
    private Paint mShadowPaint;
    private Drawable mCheckDrawable;
    private float mDensity;
    private Rect mCheckRect;
    private boolean mEnabled = true;

    public CheckView(Context context) {
        super(context);
        init(context);
    }

    public CheckView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public CheckView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // fixed size 48dp x 48dp
        int sizeSpec = MeasureSpec.makeMeasureSpec((int) (SIZE * mDensity), MeasureSpec.EXACTLY);
        super.onMeasure(sizeSpec, sizeSpec);
    }

    private void init(Context context) {
        mDensity = context.getResources().getDisplayMetrics().density;

        mStrokePaint = new Paint();
        mStrokePaint.setAntiAlias(true);
        mStrokePaint.setStyle(Paint.Style.STROKE);
        mStrokePaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_OVER));
        mStrokePaint.setStrokeWidth(STROKE_WIDTH * mDensity);
        int color = Color.WHITE;
        mStrokePaint.setColor(color);
        mCheckDrawable = ResourcesCompat.getDrawable(context.getResources(),
                R.drawable.ic_check_white_18dp, context.getTheme());
    }

    public void setChecked(boolean checked) {
        mChecked = checked;
        invalidate();
    }

    public void setEnabled(boolean enabled) {
        if (mEnabled != enabled) {
            mEnabled = enabled;
            invalidate();
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // draw outer and inner shadow
        initShadowPaint();
        canvas.drawCircle((float) SIZE * mDensity / 2, (float) SIZE * mDensity / 2,
                (STROKE_RADIUS + STROKE_WIDTH / 2 + SHADOW_WIDTH) * mDensity, mShadowPaint);

        // draw white stroke
        canvas.drawCircle((float) SIZE * mDensity / 2, (float) SIZE * mDensity / 2,
                STROKE_RADIUS * mDensity, mStrokePaint);

        // draw content
        if (mChecked) {
            initBackgroundPaint();
            canvas.drawCircle((float) SIZE * mDensity / 2, (float) SIZE * mDensity / 2,
                    BG_RADIUS * mDensity, mBackgroundPaint);

            mCheckDrawable.setBounds(getCheckRect());
            mCheckDrawable.draw(canvas);
        }

        // enable hint
        setAlpha(mEnabled ? 1.0f : 0.5f);
    }

    private void initShadowPaint() {
        if (mShadowPaint == null) {
            mShadowPaint = new Paint();
            mShadowPaint.setAntiAlias(true);
            // all in dp
            float outerRadius = STROKE_RADIUS + STROKE_WIDTH / 2;
            float innerRadius = outerRadius - STROKE_WIDTH;
            float gradientRadius = outerRadius + SHADOW_WIDTH;
            float stop0 = (innerRadius - SHADOW_WIDTH) / gradientRadius;
            float stop1 = innerRadius / gradientRadius;
            float stop2 = outerRadius / gradientRadius;
            float stop3 = 1.0f;
            mShadowPaint.setShader(
                    new RadialGradient((float) SIZE * mDensity / 2,
                            (float) SIZE * mDensity / 2,
                            gradientRadius * mDensity,
                            new int[]{Color.parseColor("#00000000"), Color.parseColor("#0D000000"),
                                    Color.parseColor("#0D000000"), Color.parseColor("#00000000")},
                            new float[]{stop0, stop1, stop2, stop3},
                            Shader.TileMode.CLAMP));
        }
    }

    private void initBackgroundPaint() {
        if (mBackgroundPaint == null) {
            mBackgroundPaint = new Paint();
            mBackgroundPaint.setAntiAlias(true);
            mBackgroundPaint.setStyle(Paint.Style.FILL);
            int color = ContextCompat.getColor(getContext(), app.melon.composer.R.color.colorPrimary);
            mBackgroundPaint.setColor(color);
        }
    }

    private void initTextPaint() {
        if (mTextPaint == null) {
            mTextPaint = new TextPaint();
            mTextPaint.setAntiAlias(true);
            mTextPaint.setColor(Color.WHITE);
            mTextPaint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));
            mTextPaint.setTextSize(12.0f * mDensity);
        }
    }

    // rect for drawing checked number or mark
    private Rect getCheckRect() {
        if (mCheckRect == null) {
            int rectPadding = (int) (SIZE * mDensity / 2 - CONTENT_SIZE * mDensity / 2);
            mCheckRect = new Rect(rectPadding, rectPadding,
                    (int) (SIZE * mDensity - rectPadding), (int) (SIZE * mDensity - rectPadding));
        }

        return mCheckRect;
    }
}

File: module_composer/src/main/java/app/melon/composer/ui/CheckView.kt
package app.melon.composer.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RadialGradient
import android.graphics.Rect
import android.graphics.Shader
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import androidx.core.content.res.ResourcesCompat
import app.melon.composer.R
import app.melon.util.extensions.dp
import app.melon.util.extensions.dpInt
import app.melon.util.extensions.getColorCompat


internal class CheckView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    var checked = false
        set(value) {
            field = value
            invalidate()
        }

    private lateinit var mStrokePaint: Paint
    private lateinit var mShadowPaint: Paint
    private lateinit var mBackgroundPaint: Paint
    private lateinit var mCheckDrawable: Drawable

    private val mSize = 48.dpInt
    private val mContentSize = 16.dpInt

    private val mStrokeWidth = 3f.dp
    private val mStrokeRadius = 11.5f.dp
    private val mShadowWidth = 6f.dp
    private val mBackgroundRadius = 11f.dp

    // rect for drawing checked number or mark
    private val checkRect: Rect by lazy {
        val rectPadding = (mSize - mContentSize) / 2
        Rect(
            rectPadding, rectPadding,
            mSize - rectPadding, mSize - rectPadding
        )
    }

    init {
        initCheckDrawable()
        initStrokePaint()
        initShadowPaint()
        initBackgroundPaint()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        // fixed size 48dp x 48dp
        val sizeSpec = MeasureSpec.makeMeasureSpec(mSize, MeasureSpec.EXACTLY)
        super.onMeasure(sizeSpec, sizeSpec)
    }

    override fun setEnabled(enabled: Boolean) {
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // draw outer and inner shadow
        canvas.drawCircle(
            mSize / 2f, mSize / 2f,
            mStrokeRadius + mStrokeWidth / 2f + mShadowWidth, mShadowPaint
        )

        // draw white stroke
        canvas.drawCircle(
            mSize / 2f, mSize / 2f,
            mStrokeRadius, mStrokePaint
        )

        // draw content
        if (checked) {
            canvas.drawCircle(
                mSize / 2f, mSize / 2f,
                mBackgroundRadius, mBackgroundPaint
            )
            mCheckDrawable.bounds = checkRect
            mCheckDrawable.draw(canvas)
        }

        // enable hint
        alpha = if (isEnabled) 1.0f else 0.5f
    }

    private fun initCheckDrawable() {
        mCheckDrawable = ResourcesCompat.getDrawable(
            context.resources,
            R.drawable.ic_check_white_18dp,
            context.theme
        )!!
    }

    private fun initStrokePaint() {
        mStrokePaint = Paint().apply {
            isAntiAlias = true
            style = Paint.Style.STROKE
            xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OVER)
            strokeWidth = mStrokeWidth
            color = Color.WHITE
        }
    }

    private fun initShadowPaint() {
        mShadowPaint = Paint()
        mShadowPaint.isAntiAlias = true
        // all in dp
        val outerRadius = mStrokeRadius + mStrokeWidth / 2
        val innerRadius = outerRadius - mStrokeWidth
        val gradientRadius = outerRadius + mShadowWidth
        val stop0 = (innerRadius - mShadowWidth) / gradientRadius
        val stop1 = innerRadius / gradientRadius
        val stop2 = outerRadius / gradientRadius
        val stop3 = 1.0f
        mShadowPaint.shader = RadialGradient(
            mSize / 2f,
            mSize / 2f,
            gradientRadius, intArrayOf(
                Color.parseColor("#00000000"), Color.parseColor("#0D000000"),
                Color.parseColor("#0D000000"), Color.parseColor("#00000000")
            ), floatArrayOf(stop0, stop1, stop2, stop3),
            Shader.TileMode.CLAMP
        )
    }

    private fun initBackgroundPaint() {
        mBackgroundPaint = Paint().apply {
            isAntiAlias = true
            style = Paint.Style.FILL
            color = context.getColorCompat(R.color.colorPrimary)
        }
    }
}